Lib/Src/adc_data.cpp

```diff
@@ -0,0 +1,95 @@
+/**
+ *@file adc_data.cpp
+ *@brief This files contains the methods for the CAdc class
+ *
+ */
+
+/*
+ *  Created on: Feb 14, 2022
+ *      Author: niuslar
+ */
+
+#include "adc_data.h"
+
+#define ADC_START_BIT   2
+#define ALARM_TOGGLE_MS 300
+
+CUartCom uart_adc(&huart2);
+
+CLog log_adc(&uart_adc, "ADC_data");
+
+/**
+ * @brief Calibrate and Start ADC with DMA
+ * @note The ADC will read all the channels once and then stop.
+ * To read the channels again, use the triggerADC() method.
+ * @param p_hadc Pointer to ADC handler
+ * @param adc_channels Number of active channels for the ADC
+ * 		(This depends on the hardware configuration)
+ */
+CAdc::CAdc(ADC_HandleTypeDef *p_hadc, uint8_t adc_channels)
+    : mp_hadc(p_hadc),
+      m_adc_channels(adc_channels)
+{
+    // Check mp_hadc is not null
+    if (!mp_hadc)
+    {
+        // Send error message
+        log_adc.log(CLog::LOG_ERROR, "Invalid pointer to ADC Handler");
+    }
+
+    // Allocate memory to store the incoming data
+    mp_adc_data_buf = new uint16_t[adc_channels];
+
+    // Calibrate ADC
+    if (HAL_ADCEx_Calibration_Start(p_hadc, ADC_SINGLE_ENDED) != HAL_OK)
```

Comment:
This will fail. Constructors are executed when classes are created, which will happen before anything else is executed. You constructor should do basic checks and that's it. No calls to hardware at all. Then you create a special `init()` method that you run to do all of this fancy stuff.

Reply:
Also, make sure your constructor leaves the class in a safe state in case init() is never executed.